### DISCUSSION TITLE
AI model performance

### DISCUSSION THREAD

**Discussion by OOHehir** [LIFsCode/ELOC-3.0#99] [General]
@EDsteve @LIFsCode Rather than having this discussed over 'issues' I thought it better to start a discussion. 

I've added some [test code](eloc610LowPowerPartition/test/test_target_ai_model/test.cpp) that uses the project libraries but otherwise avoids disrupting the 'production' code. It runs any wav files found in the SD card (/sdcard/wav_test_files/) though the project ai model (found in lib/edge-impulse).

Where the wav file sample rate is greater than the model sample rate it applies the current method of down-sampling. It appears to give identical results, a little suspicious but I've yet to find a reason that these results can't be believed.  E.g. sample Trumpet2.wav run through the 4K model:

4K 0.996094 
8K 0.996094 
16K 0.996094 

Here's some other results run through the current model:

```
I (18846) getFileListWithExtension: Found filename: 8K_Other1.wav
I (18847) getFileListWithExtension: Found filename: 8K_Other2.wav
I (18849) getFileListWithExtension: Found filename: 8K_Trumpet1.wav
I (18856) getFileListWithExtension: Found filename: 8K_Trumpet2.wav
I (18863) getFileListWithExtension: Found filename: 16K_Oother2.wav
I (18871) getFileListWithExtension: Found filename: 16K_Other1.wav
I (18877) getFileListWithExtension: Found filename: 16K_Trumpet1.wav
I (18884) getFileListWithExtension: Found filename: 16K_Trumpet2.wav
I (18891) getFileListWithExtension: Found filename: 4k_Oother2.wav
I (18897) getFileListWithExtension: Found filename: 4k_Other1.wav
I (18904) getFileListWithExtension: Found filename: 4k_Trumpet1.wav
I (18912) getFileListWithExtension: Found filename: 4k_Trumpet2.wav
Testing file /sdcard/wav_test_files/8K_Other1.wav 
I (18929) WAV: num_channels=1, sample_rate=8000, bit_depth=16, number of samples=8001
Test model predictions: 
    (DSP: 242 ms., Classification: 44 ms., Anomaly: 0 ms.) 
    background: 0.179688 
    trumpet: 0.820312 
Testing file /sdcard/wav_test_files/8K_Other2.wav 
I (19601) WAV: num_channels=1, sample_rate=8000, bit_depth=16, number of samples=8001
Test model predictions: 
    (DSP: 242 ms., Classification: 44 ms., Anomaly: 0 ms.) 
    background: 0.996094 
    trumpet: 0.000000 
Testing file /sdcard/wav_test_files/8K_Trumpet1.wav 
I (20273) WAV: num_channels=1, sample_rate=8000, bit_depth=16, number of samples=8001
Test model predictions: 
    (DSP: 242 ms., Classification: 44 ms., Anomaly: 0 ms.) 
    background: 0.003906 
    trumpet: 0.996094 
Testing file /sdcard/wav_test_files/8K_Trumpet2.wav 
I (20945) WAV: num_channels=1, sample_rate=8000, bit_depth=16, number of samples=8001
Test model predictions: 
    (DSP: 242 ms., Classification: 44 ms., Anomaly: 0 ms.) 
    background: 0.000000 
    trumpet: 0.996094 
Testing file /sdcard/wav_test_files/16K_Oother2.wav 
I (21617) WAV: num_channels=1, sample_rate=16000, bit_depth=16, number of samples=16001
Test model predictions: 
    (DSP: 242 ms., Classification: 44 ms., Anomaly: 0 ms.) 
    background: 0.996094 
    trumpet: 0.000000 
Testing file /sdcard/wav_test_files/16K_Other1.wav 
I (22557) WAV: num_channels=1, sample_rate=16000, bit_depth=16, number of samples=16001
Test model predictions: 
    (DSP: 242 ms., Classification: 44 ms., Anomaly: 0 ms.) 
    background: 0.996094 
    trumpet: 0.000000 
Testing file /sdcard/wav_test_files/16K_Trumpet1.wav 
I (23496) WAV: num_channels=1, sample_rate=16000, bit_depth=16, number of samples=16001
Test model predictions: 
    (DSP: 242 ms., Classification: 44 ms., Anomaly: 0 ms.) 
    background: 0.031250 
    trumpet: 0.968750 
Testing file /sdcard/wav_test_files/16K_Trumpet2.wav 
I (24436) WAV: num_channels=1, sample_rate=16000, bit_depth=16, number of samples=16001
Test model predictions: 
    (DSP: 242 ms., Classification: 44 ms., Anomaly: 0 ms.) 
    background: 0.000000 
    trumpet: 0.996094 
Testing file /sdcard/wav_test_files/4k_Oother2.wav 
I (25375) WAV: num_channels=1, sample_rate=4000, bit_depth=16, number of samples=4000
Test model predictions: 
    (DSP: 242 ms., Classification: 44 ms., Anomaly: 0 ms.) 
    background: 0.996094 
    trumpet: 0.000000 
Testing file /sdcard/wav_test_files/4k_Other1.wav 
I (25908) WAV: num_channels=1, sample_rate=4000, bit_depth=16, number of samples=4000
Test model predictions: 
    (DSP: 242 ms., Classification: 44 ms., Anomaly: 0 ms.) 
    background: 0.968750 
    trumpet: 0.031250 
Testing file /sdcard/wav_test_files/4k_Trumpet1.wav 
I (26441) WAV: num_channels=1, sample_rate=4000, bit_depth=16, number of samples=4000
Test model predictions: 
    (DSP: 242 ms., Classification: 44 ms., Anomaly: 0 ms.) 
    background: 0.000000 
    trumpet: 0.996094 
Testing file /sdcard/wav_test_files/4k_Trumpet2.wav 
I (26976) WAV: num_channels=1, sample_rate=4000, bit_depth=16, number of samples=4000
Test model predictions: 
    (DSP: 242 ms., Classification: 44 ms., Anomaly: 0 ms.) 
    background: 0.000000 
    trumpet: 0.996094 
```

**Comment by OOHehir** [comment]
@EDsteve I understand you're getting Guru errors. Do you have an image of the SD contents?

**Reply by EDsteve** [reply]
That's the folder structure and files from the SD card:
![image](https://github.com/LIFsCode/ELOC-3.0/assets/25384076/d0f7743f-4a77-47f5-80ad-45ab4b4e0c6e)

**Reply by OOHehir** [reply]
@EDsteve Apologies, I believe I found the issue. Before the test is run that checks all available wav files it runs another test on specific files that you don't have in your SD card. I'm testing a fix now, should be ready shortly..

**Reply by OOHehir** [reply]
@EDsteve If you update the code your setup should work now.

**Reply by EDsteve** [reply]
@OOHehir 
Awesome. It works now. Time to test some models on different sound files :)

**Comment by EDsteve** [comment]
@OOHehir 
I don't think you do right now. But just in case you play with AI training. This info might be crucial. 
I just found out that our database admin saved the trimmed features as 24bit sounds. But the original sound files and the ELOC uses 16bit. So we trained our model with 24bit sound files. That could be a reason why the model does not perform well on the 16bit ELOC?
I will find out after i converted the files to 16bit and retrained :)

**Reply by OOHehir** [reply]
@EDsteve Certainly sounds like a reason the model wouldn't perform well on the ELOC. Thanks for the info.

**Reply by EDsteve** [reply]
Background sound was 16bit and features 24bit. That can't be good. Haha

**Comment by EDsteve** [comment]
@OOHehir 
I have a new model i want to test (trained with 16bit files), but getting build errors again. Is there something i can do to prevent that? Because i am sure i will try many models and don't want to annoy you all the time about it :)
(I have added you as a collaborator in EI, in case you want to take a look) 

**Done so far:**
- This line is uncommented: #define EI_BUFFER_IN_PSRAM
- I have tried master and model_290754_68 branch
- line 120 is correctly modified

Here the new models (trained with 16bit) i tried to integrate, which outperformed all other models on EI (using life classification). So time to test it on the ELOC. 
[ei-trumpets-16bit-arduino-1.0.11.zip](https://github.com/LIFsCode/ELOC-3.0/files/15201744/ei-trumpets-16bit-arduino-1.0.11.zip)
[ei-trumpets-16bit-arduino-1.0.10.zip](https://github.com/LIFsCode/ELOC-3.0/files/15201739/ei-trumpets-16bit-arduino-1.0.10.zip)

Here the build error:
```
Compiling .pio\build\esp32dev-ei-windows\lib882\edge-impulse\edge-impulse-sdk\CMSIS\DSP\Source\StatisticsFunctions\arm_absmin_no_idx_f16.c.o
In file included from lib/edge-impulse/src/edge-impulse-sdk/classifier/ei_run_dsp.h:25,
                 from lib/edge-impulse/src/edge-impulse-sdk/classifier/ei_run_classifier.h:24,
                 from lib/edge-impulse/src/trumpets_16bit_inferencing.h:49,
                 from lib/edge-impulse/src/EdgeImpulse.cpp:10:
lib/edge-impulse/src/edge-impulse-sdk/dsp/ei_flatten.h: In member function 'virtual int flatten_class::extract(ei::signal_t*, ei::matrix_t*, void*, float)':
lib/edge-impulse/src/edge-impulse-sdk/dsp/ei_flatten.h:183:31: error: 'mean' may be used uninitialized in this function [-Werror=maybe-uninitialized]
             means[axis][head] = mean;
lib/edge-impulse/src/edge-impulse-sdk/dsp/ei_flatten.h:90:19: note: 'mean' was declared here
             float mean; // to use with moving average
                   ^~~~
Compiling .pio\build\esp32dev-ei-windows\lib882\edge-impulse\edge-impulse-sdk\CMSIS\DSP\Source\StatisticsFunctions\arm_absmin_no_idx_f32.c.o
Compiling .pio\build\esp32dev-ei-windows\lib882\edge-impulse\edge-impulse-sdk\CMSIS\DSP\Source\StatisticsFunctions\arm_kullback_leibler_f16.c.o
lib/edge-impulse/src/edge-impulse-sdk/dsp/ei_flatten.h: In function 'EI_IMPULSE_ERROR {anonymous}::process_impulse(ei_impulse_handle_t*, ei::signal_t*, ei_impulse_result_t*, bool)':
lib/edge-impulse/src/edge-impulse-sdk/dsp/ei_flatten.h:183:31: error: 'mean' may be used uninitialized in this function [-Werror=maybe-uninitialized]
             means[axis][head] = mean;
lib/edge-impulse/src/edge-impulse-sdk/dsp/ei_flatten.h:90:19: note: 'mean' was declared here
             float mean; // to use with moving average
                   ^~~~
Compiling .pio\build\esp32dev-ei-windows\lib882\edge-impulse\edge-impulse-sdk\CMSIS\DSP\Source\StatisticsFunctions\arm_kullback_leibler_f32.c.o
Compiling .pio\build\esp32dev-ei
```

**Comment by OOHehir** [comment]
@EDsteve You'll probably see alot of these if you're changing model. The quick solution is to change the line:

```
lib/edge-impulse/src/edge-impulse-sdk/dsp/ei_flatten.h:90:19: note: 'mean' was declared here
             float mean; // to use with moving average
```
to
```
lib/edge-impulse/src/edge-impulse-sdk/dsp/ei_flatten.h:90:19: note: 'mean' was declared here
             float mean = 0; // to use with moving average
```
The reason for this is that the compiler is set to error on these warning. This setting can be tailored per folder with something like this [feature](https://docs.platformio.org/en/latest/scripting/middlewares.html#build-middlewares). Will take a few days to get it implemented.

**Reply by EDsteve** [reply]
Thanks for the info. I can work with that.

**Comment by OOHehir** [comment]
@EDsteve This [commit](https://github.com/LIFsCode/ELOC-3.0/commit/2bb324a47c73ecbc55fa8f5471f91ec8b975041a) should hopefully solve the issue, please let me know if it doesn't work.

**Comment by EDsteve** [comment]
@OOHehir 
Even with the last commit. I still had to change in code to `float mean = 0;` But not a problem.

I have made a model with a window size of 2 seconds which worked well on EI and my laptop. But didn't work at all on the ELOC with 2 seconds test wav files. Gave me totally wrong predictions.
I thought it's a problem with the window size of 2 seconds. So i made a new model with 1sec. This model seems to outperform everything i had before. It predicts over 90% of unseen trumpets. Predicts almosst everything correctly when i deploy it on my laptop or on my two phones. So that should work also on the ELOC. Theoretically.

Unfortunately it gives me an error about having not enough RAM, which i can't really understand. (I uncommented the line  #define EI_BUFFER_IN_PSRAM). We have 8MB PSRAM. How can that be full? According to EI the model only needs a few KB:
![image](https://github.com/LIFsCode/ELOC-3.0/assets/25384076/b6d92659-0dd3-4e79-8b8a-d6ce7e445b60)

If you could take a look that would be great.

Here is the model with already modified select.cpp:
[ei-trumpets-16bit-arduino-1.0.17.zip](https://github.com/LIFsCode/ELOC-3.0/files/15233486/ei-trumpets-16bit-arduino-1.0.17.zip)

```
Linking .pio\build\target_unit_selected_tests\firmware.elf
Retrieving maximum program size .pio\build\target_unit_selected_tests\firmware.elf
Checking size .pio\build\target_unit_selected_tests\firmware.elf
Advanced Memory Usage is available via "PlatformIO Home > Project Inspect"
RAM:   [===       ]  26.7% (used 87516 bytes from 327680 bytes)
Error: The program size (1502097 bytes) is greater than maximum allowed (1048576 bytes)
Flash: [====*** [checkprogsize] Explicit exit, status 1
======]  143.3% (used 1502097 bytes from 1048576 bytes)
```

**Comment by OOHehir** [comment]
@EDsteve

I've modified the platformio.ini file, it should now build models without errors.

I've run the esp32dev-ei build & it looks fine:
```
Checking size .pio/build/esp32dev-ei/firmware.elf
Advanced Memory Usage is available via "PlatformIO Home > Project Inspect"
RAM:   [====      ]  35.7% (used 116932 bytes from 327680 bytes)
Flash: [==        ]  20.7% (used 1712673 bytes from 8257536 bytes)
Building .pio/build/esp32dev-ei/firmware.bin
esptool.py v4.5
Creating esp32 image...
Merged 25 ELF sections
```

When your testing the wav files please make sure you using Advanced > Verbose Test

This task basically replaces the usual main.cpp with a test.cpp. Doing a 'regular' build will probably end up pulling in both the regular & test project.

BTW. The latest commit also pulls in the model you included above.

**Comment by EDsteve** [comment]
@OOHehir 
All right. We are getting closer and closer :)

For testing I have selected four different trumpets (unseen by the model). Each has a length of one second. 
[test_trumpets_NEW.zip](https://github.com/LIFsCode/ELOC-3.0/files/15245588/test_trumpets_NEW.zip)
Trumpet 1: very easy
Trumpet 2, 3 and 4: a bit more difficult

**Test results:**
- When i feed these trumpets to the model in EI (using live classification), three out of 4 gets predicted correctly with 0.99 or higher accuracy. And one gets a 0.37 prediction. 
- When i do the Verbose Test on the ELOC with the same sound files. Only one trumpet (the easiest) gets recognized (0.99).

Using wav files. The model on the ELOC should have (almost) the same outcome compared with the model on EI, correct? I think that's what we should get working first before i start playing with inference using the microphone.

I understand that I2S_DEFAULT_VOLUME adjusts the volume/gain for the I2S. Does something similar exist with how the WAV file gets fed to the model? Or do you see another reason, why the model on the ELOC predicts very differently compared to EI live classification?

This is the model in case you want to check live classification: https://studio.edgeimpulse.com/studio/243836


ELOC Verbose Test results:
```
Testing file /sdcard/wav_test_files/1sec_NEW_trumpet1.wav 
I (11238) WAV: num_channels=1, sample_rate=16000, bit_depth=16, number of samples=16001
Skip rate = 4
Test model predictions: 
    (DSP: 478 ms., Classification: 183 ms., Anomaly: 0 ms.)
    others: 0.000000
    trumpet: 0.996094
Testing file /sdcard/wav_test_files/1sec_NEW_trumpet2.wav 
I (12543) WAV: num_channels=1, sample_rate=16000, bit_depth=16, number of samples=16000
Skip rate = 4
Test model predictions: 
    (DSP: 478 ms., Classification: 183 ms., Anomaly: 0 ms.)
    others: 0.996094
    trumpet: 0.000000
Testing file /sdcard/wav_test_files/1sec_NEW_trumpet3.wav 
I (13849) WAV: num_channels=1, sample_rate=16000, bit_depth=16, number of samples=16001
Skip rate = 4
Test model predictions: 
    (DSP: 478 ms., Classification: 183 ms., Anomaly: 0 ms.)
    others: 0.996094
    trumpet: 0.003906
Testing file /sdcard/wav_test_files/1sec_NEW_trumpet4.wav 
I (15155) WAV: num_channels=1, sample_rate=16000, bit_depth=16, number of samples=16001
Skip rate = 4
Test model predictions: 
    (DSP: 478 ms., Classification: 183 ms., Anomaly: 0 ms.)
    others: 0.996094
    trumpet: 0.000000
test/test_target_ai_model/test.cpp:296:test_ai_model:PASS
```

Looking at the numbers. Is there are reason why it never predicts with 1.00? It seems the highest number is 0.996094. I just want to point it out. No clue if that is normal or not. :)

**Comment by OOHehir** [comment]
@EDsteve 

_When i feed these trumpets to the model in EI (using live classification), three out of 4 gets predicted correctly with 0.99 or higher accuracy. And one gets a 0.37 prediction._

That maybe a sign that the model is [overfitting](www.ibm.com/topics/overfitting) the data. I'd be worried if a model reached a prediction value of 1.0, it probably means it has been tested on data that it's been trained on. I think a reasonable expectation would be 0.8 or greater. If you aim for too high a prediction value it means the model won't generalize well & will struggle to classify trumpets that aren't very close to the training data.

_When i do the Verbose Test on the ELOC with the same sound files. Only one trumpet (the easiest) gets recognized (0.99)._

This could be a sign that the downsampling technique is affecting the model performance. I think to have confidence in the model performance you'll always need to include wav files samples at 4k, 8k & 16k. Significant difference in the performance will indicate that downsampling is not suitable for the model.

I'm not certain, but suspect, that if a model has an 'averaging' (i.e. pooling layer) as it's input stage it's less  likekly be be affected by the downsampling.

_I understand that I2S_DEFAULT_VOLUME adjusts the volume/gain for the I2S. Does something similar exist with how the WAV file gets fed to the model?_

No, the wav file gets fed straight into the model. The reason for the volume adjustment was to handle the samples that exceed the 16 bit value (as the I2S mic produces 24 bit samples).  I'm fairly sure any sound conversion process will cater for this also. If the 16 bit wav file sounds OK on playback it should be fine.

**Comment by EDsteve** [comment]
@OOHehir Very cool. Your tip with the 4K sound files made all the difference. Using 4K wav files works great and i get similar results compared to the EI Live Classification. That makes me so happy now :)
As before i used EI to downsample the test files and then copied it to the ELOC. But that also means that the downsampling on the ELOC is different compared to the downsampling on the ELOC? 
Does the downsampling for AI and for recording work the same way? If yes. I can compare 4K recordings from the ELOC with the 4K samples of EI.

Good point with the overfitting. I will make some adjustments (Drop out layer, Epochs, learning rate, data augmentation,..) and try to get predictions not as "sharp" as they are now. And also report back with results.

### Here some results
**16K wav files (on ELOC Verbose test):**
| Sound File                | Others |  Trumpet |
|---------------------------|------------------------|-------------------------|
| 1sec_NEW_trumpet1.wav     | 0.000000               | 0.996094                |
| 1sec_NEW_trumpet2.wav     | 0.996094               | 0.000000                |
| 1sec_NEW_trumpet3.wav     | 0.996094               | 0.003906                |
| 1sec_NEW_trumpet4.wav     | 0.996094               | 0.000000                |  


**4K wav files (on ELOC Verbose test):**
| Sound File                  | Others | Trumpet |
|-----------------------------|------------------------|-------------------------|
| 1sec_NEW_4K_trumpet1.wav    | 0.000000               | 0.996094                |
| 1sec_NEW_4K_trumpet2.wav    | 0.718750               | 0.281250                |
| 1sec_NEW_4K_trumpet3.wav    | 0.000000               | 0.996094                |
| 1sec_NEW_4K_trumpet4.wav    | 0.015625               | 0.984375                |

**Edge Impulse Live Clsssification:**
| Sound File                | Others |  Trumpet |
|---------------------------|------------------------|-------------------------|
| 1sec_NEW_trumpet1.wav     | 0.00               | 1.00                |
| 1sec_NEW_trumpet2.wav     | 0.63               | 0.37                |
| 1sec_NEW_trumpet3.wav     | 0.00               | 1.00               |
| 1sec_NEW_trumpet4.wav     | 0.01               | 0.99                |  

All wav files used have not been seen by the model during training.

**Reply by OOHehir** [reply]
@EDsteve Yes, it looks like the 4k results are very similar to the EI live classification so you can have confidence in the process on the ELOC.

As discussed before, the downsampling on the ELOC is very crude, it simply drops samples. To be done properly, an FFT filter would probably be needed. I've had a quick look & didn't find a simple solution. A separate codec chip might be required or alot of computation, at which point (perhaps) you might just be better using a 16k model. It's certainly a tradeoff.

**Comment by EDsteve** [comment]
@OOHehir 
Okay. Understood. In the constelation we have right now. Propper downsampling is not worth it. Either make the model good enough to "read" the 4K files or just use 16K.

Another thing i realized is that the DSP and inference takes much longer on the ELOC as predicted on EI. 
EI stated 136ms for DSP on the ESP32. But running it on the ELOC it takes 478ms.
Same with Inference. EI predicted 80ms but on the ELOC it takes 183ms.

Is EI just wrong or does the model not get all the CPU power of the ESP32? Models work best with a window increase size of around 500ms.So DSP and inference time of more than 500ms is problematic i guess?

```
Testing file /sdcard/wav_test_files/1sec_NEW_4K_trumpet4.wav 
␛[0;32mI (19275) WAV: num_channels=1, sample_rate=4000, bit_depth=16, number of samples=8000␛[0m
Test model predictions: 
    (DSP: 478 ms., Classification: 183 ms., Anomaly: 0 ms.)
    others: 0.015625
    trumpet: 0.984375
```

**Reply by OOHehir** [reply]
@EDsteve If you take a look at the device in EI it shows an ESP32 running at, I think, 240Mhz. For power saving the ELOC is run at, I think, 80Mhz. I believe that's the cause.

**Comment by EDsteve** [comment]
I know i am hijacking this thread for AI improvements. Hope thats fine :)
@OOHehir

I have another idea to make a solid model which might be much better with 4K files:

As you mentioned. We could set the model to train on 16K files (the model right now is for 16K). But that comes with the cost of higher resources.
What if we convert the original 16K trumpet sound files to 4K sound files using the same procedure as the ELOC does (just reducing some samples without filter, right?). And then train the model with these new files. The resulting model should be super accurate running inference with 4K on the ELOC? And also reduces resources. 

I won't have time for it before i send the ELOCs, but will definitely try the next weeks. 
my plan is to ask chatGPT: Write s script which reduces the bit rate from all files in this folder to 4K in the same way as this code does: "your code from ELOC"
Can you link me to the code which converts the I2S stream to a lower bit rate? So i can feed it to GPT.

**Comment by OOHehir** [comment]
@EDsteve That's a great idea! I wonder how the frequency of the sound file might be distorted, but perhaps the AI model will cope. It's certainly worth a try. The code is fairly straightforward & located [here](https://github.com/LIFsCode/ELOC-3.0/blob/7cc2678882dd37407c452bd3d710b98363989a4a/eloc610LowPowerPartition/test/test_target_edge-impulse/test.cpp#L110). If you've any issues with generating the code let me know & I can take a look.

**Comment by LIFsCode** [comment]
@OOHehir @EDsteve 
I know it is not really related to this issue, but I don't know where else to put as it is a very minor thing:

1. Is the content of the [edge-impulse-sdk folder](https://github.com/LIFsCode/ELOC-3.0/tree/master/eloc610LowPowerPartition/lib/edge-impulse/src/edge-impulse-sdk) fully created by Edge Impulse without modifications? or are there any custom modifications in there?
2. The [EI_PORTING_ARDUINO](https://github.com/LIFsCode/ELOC-3.0/blob/0e3c1a72be40a7138100cb45f67fb5f615fd23fc/eloc610LowPowerPartition/lib/edge-impulse/src/edge-impulse-sdk/porting/ei_classifier_porting.h#L285) is redefined here? Is there any meaning it that? or should it just be forced to 0. Should we undefine it first to avoid warnings or at least ignore this warning?

**Comment by EDsteve** [comment]
@LIFsCode 
1. There are some minor modifications necessary. More info in the [Readme](https://github.com/LIFsCode/ELOC-3.0/blob/master/eloc610LowPowerPartition/README-ai.md). 
    A quick summary:
   - `/lib/edge-impulse/src/EdgeImpulse.cppn.cpp` Line 10 needs the name of the new model
   - `lib/edge-impulse/src/edge-impulse-sdk/tensorflow/lite/micro/kernels/select.cpp` Line 125 must be changed to `TfLiteIntArray* output_size = nullptr;`
 2. Hope @OOHehir still remembers :)

**Comment by OOHehir** [comment]
@LIFsCode 

1. As @EDsteve  pointed out EdgeImpulse.cpp & .h are custom & need to be updated as the model updates.
2. The build flags and defines are Edge Impulses wrapper around the TensorFlow build. Depending on the model different warnings are generated. I previously had more strict 'project wide' compiler flags but these had to be removed to allow the EI files to compile, not ideal. The TensorFlow build has many issues around un-initialised variables. I'd expect both the TensorFlow & EI code to be updated/ modified with each new model & these warning to evolve/ change.
